File: java/pilha_dinamica/PilhaDinamica.java
package pilha_dinamica;

import java.util.List;
import main.Celula;

public class PilhaDinamica {
    private int qtd;
    private Celula topo;
    
    public PilhaDinamica() {
        qtd = 0;
        topo = null;
    }

    public void empilhar(int e) {
        topo = new Celula(e, topo);
        qtd++;
    }

	public void empilharList(List<Integer> lista) {
		for (int item : lista) 
			empilhar(item);
	}

    public int desempilhar() {
        int retorno = topo.item;
        Celula tmp = topo;
        topo = topo.prox;
        tmp.prox = null;
        tmp = null;
        qtd--;
        return retorno;
    }

    public void mostrarPilha() {
        for (Celula c = topo; c != null; c = c.prox) 
            System.out.println("|\t"+c.item+"\t|");
        System.out.println();
    }
}

File: c/pilha_dinamica/pilhaDinamica.c
#include "pilhaDinamica.h"
#include "../common/celula.h"

#include <stdio.h>
#include <stdlib.h>

Pilha * newPilha() {
	Pilha * pilha = (Pilha *) malloc(sizeof(Pilha));
	pilha->topo = newCelula();
	return pilha;
}

int getTamanhoDaPilha(Pilha * pilha) {
	int tam = 0;
	for (Celula * tmp = pilha->topo; tmp != NULL; tmp = tmp->prox, tam++);
	return tam;
}

int empilhar(Pilha * pilha, int item) {
	pilha->topo = newCelula3(pilha->topo, item);
	return 1;
}

int desempilhar(Pilha * pilha) {
	int removido = -9999;
	if (pilha->topo == NULL) {
		printf("Erro ao desempilhar: pilha vazia.\n");
	} else {
		Celula * tmp = pilha->topo;
		pilha->topo = tmp->prox;
		removido = tmp->item;
		free(tmp);
	}
	return removido;
}

void mostrarItensDaPilha(Pilha * pilha) {
	for (Celula * tmp = pilha->topo; tmp != NULL; tmp = tmp->prox) {
		printf("\t|\t%d\t|\n", tmp->item);
	}
}
